ARC: Add virtual uart over ARC hostlink channel support

evgeniy-paltsev opened the pull request:
Add support for virtual UART device that uses ARC Hostlink channels for data transfers. Due to the Hostlink principle, the driver    supports only polling API.

Hostlink is supported on nSIM and HW targets (if launched with MDB debugger).

evgeniy-paltsev committed:
drivers: serial: add virtual uart over ARC hostlink channel

Add support for virtual UART device that uses ARC Hostlink channels
for data transfers. Due to the Hostlink principle, this driver
supports only polling API.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev committed:
ARC: boards: add nsim-based board with hostlink uart support

Add nsim_hs3x_hostlink nsim-based board to test virtual UART
emulated with hostlink channel.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

ruuddw commented on drivers/serial/uart_hostlink.c:
```diff
@@ -0,0 +1,410 @@
+	volatile union payload_u payload;
+} __aligned(HL_MAX_DCACHE_LINE) __packed;
+
+/* In general we mast exact fit into one or multiple cache lines as we shouldn't share hostlink
```
mast->must

evgeniy-paltsev replied:
fixed

ruuddw commented on drivers/serial/uart_hostlink.c:
```diff
@@ -0,0 +1,410 @@
+{
+	ARG_UNUSED(dev);
+
+	/* We plan to use hostlink for logging, so no much sense in poll_in implementation */
```
If it is not too complex, why not add input? That enables shell functionality, and other simple interactivity if needed.

evgeniy-paltsev replied:
As I remember shell require interrupt functionality from UART, so implementing of polling is not enough. But I'll check if it is so.

ruuddw replied:
Agreed - hostlink stdin may be blocking instead of polling, let's continue with this first.

ruuddw commented:
Great addition for bring-up and debugging. Maybe add a short description in the board documentation? Anything special commands required to start a debugger session that should be documented, like enable/disable hostlink in the debugger?

evgeniy-paltsev committed:
west: runners: mdb-hw: add hostlink-awareness

Add hostlink-awareness to mdb-hw runner. The mdb-nsim and arc-nsim
runners (as well as cmake scripting for nSIM boards) doesn't require
any changes.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>
Signed-off-by: Evgeniy Paltsev <PaltsevEvgeniy@gmail.com>

evgeniy-paltsev commented:
> Anything special commands required to start a debugger session that should be documented, like enable/disable hostlink in the debugger?

I've added hostlink-awareness to west runners - so no changes required if we start debugging session or just do with `mdb-hw`, `mdb-nsim`, `arc-nsim` west runners or with cmake scripting on nSIM.

mbolivar-nordic reviewed (commented):
Late to the party, but runner changes LGTM
